File: shared/src/androidMain/kotlin/com/luckyfrog/quickmart/core/app/AppPreferences.kt
package com.luckyfrog.quickmart.core.app

import android.content.Context
import android.util.Log
import com.luckyfrog.quickmart.core.di.provideSettings
import com.luckyfrog.quickmart.core.preferences.StringSettingConfig
import com.luckyfrog.quickmart.utils.Constants.ACCESS_TOKEN
import com.luckyfrog.quickmart.utils.Constants.APP_THEME
import com.luckyfrog.quickmart.utils.Constants.FIRST_TIME
import com.luckyfrog.quickmart.utils.Constants.REFRESH_TOKEN
import com.luckyfrog.quickmart.utils.resource.theme.AppTheme
import com.russhwolf.settings.Settings
import com.russhwolf.settings.get

object AppPreferences {
    private lateinit var settings: Settings

    fun setTheme(value: AppTheme, context: Context): Boolean {
        settings = provideSettings(context)
        settings.putString(APP_THEME, value.name)
        return (settings.get<String>(APP_THEME) ?: AppTheme.Default.name) == value.name
    }

    fun getTheme(context: Context): AppTheme {
        settings = provideSettings(context)
        settings.putString(APP_THEME, AppTheme.Default.name)
        return AppTheme.valueOf(settings.get<String>(APP_THEME) ?: AppTheme.Default.name)
    }

    fun setFirstTime(context: Context): String {
        settings = provideSettings(context)
        val config =
            StringSettingConfig(settings, FIRST_TIME, "")
        config.set("1")
        Log.i("AppPreferences", "getFirstTime: ${config.get()}")

        return config.get()
    }

    fun getFirstTime(context: Context): String {
        settings = provideSettings(context)
        val config =
            StringSettingConfig(settings, FIRST_TIME, "")
        return config.get()
    }

    fun setToken(value: String, context: Context): String {
        settings = provideSettings(context)
        val configAccessToken =
            StringSettingConfig(settings, ACCESS_TOKEN, "")
        configAccessToken.set(value)
        return configAccessToken.get()
    }

    fun getToken(context: Context): String {
        settings = provideSettings(context)
        val config =
            StringSettingConfig(settings, ACCESS_TOKEN, "")
        return config.get()
    }

    fun setRefreshToken(value: String, context: Context): String {
        settings = provideSettings(context)
        val config =
            StringSettingConfig(settings, REFRESH_TOKEN, "")
        config.set(value)
        return config.get()
    }

    fun getRefreshToken(context: Context): String {
        settings = provideSettings(context)
        val configAccessToken =
            StringSettingConfig(settings, REFRESH_TOKEN, "")
        return configAccessToken.get()
    }

    fun clearToken(context: Context): Boolean {
        settings = provideSettings(context)
        val token =
            StringSettingConfig(settings, ACCESS_TOKEN, "")
        val refreshToken =
            StringSettingConfig(settings, REFRESH_TOKEN, "")
        token.remove()
        refreshToken.remove()
        return true
    }
}

File: iosApp/iosApp/Utils/AppPreferences.swift
//
//  Settings.swift
//  iosApp
//
//  Created by Eric on 17/12/24.
//  Copyright © 2024 orgName. All rights reserved.
//
import Foundation
import SwiftUI
import Shared

class StringSettingConfig {
    private let settings: Settings
    private let key: String
    private let defaultValue: String
    
    init(settings: Settings, key: String, defaultValue: String) {
        self.settings = settings
        self.key = key
        self.defaultValue = defaultValue
    }
    
    func remove() {
        settings.remove(key: key)
    }
    
    func exists() -> Bool {
        return settings.hasKey(key: key)
    }
    
    func get() -> String {
        return settings.getString(key: key, defaultValue: defaultValue)
    }
    
    func set(_ value: String) -> Bool {
        do {
            settings.putString(key: key, value: value)
            return true
        } catch {
            return false
        }
    }
}

class AppPreferences {
    private var settings: Settings
    
    // Singleton instance
    static let shared = AppPreferences()
    
    // Private initializer to ensure singleton pattern
    private init() {
        // This is where you'd use NSUserDefaults settings
        settings = NSUserDefaultsSettings(delegate: UserDefaults.standard)
    }
    
    // Theme Management
    func setTheme(_ value: AppTheme) -> Bool {
        settings.putString(key: "app_theme", value: value.rawValue)
        return (settings.getString(key: "app_theme", defaultValue: AppTheme.default.rawValue)) == value.rawValue
    }
    
    func getTheme() -> AppTheme {
        settings.putString(key: "app_theme", value: AppTheme.default.rawValue)
        return AppTheme(rawValue: settings.getString(key: "app_theme", defaultValue: AppTheme.default.rawValue)) ?? .default
    }
    
    // First Time Flag
    func setFirstTime(_ value: String = "1") -> String {
        let config = StringSettingConfig(settings: settings, key: "first_time", defaultValue: "")
        _ = config.set(value)
        return config.get()
    }
    
    func getFirstTime() -> String {
        let config = StringSettingConfig(settings: settings, key: "first_time", defaultValue: "")
        return config.get()
    }
    
    // Access Token Management
    func setToken(_ value: String) -> String {
        let config = StringSettingConfig(settings: settings, key: "access_token", defaultValue: "")
        _ = config.set(value)
        return config.get()
    }
    
    func getToken() -> String {
        let config = StringSettingConfig(settings: settings, key: "access_token", defaultValue: "")
        return config.get()
    }
    
    // Refresh Token Management
    func setRefreshToken(_ value: String) -> String {
        let config = StringSettingConfig(settings: settings, key: "refresh_token", defaultValue: "")
        _ = config.set(value)
        return config.get()
    }
    
    func getRefreshToken() -> String {
        let config = StringSettingConfig(settings: settings, key: "refresh_token", defaultValue: "")
        return config.get()
    }
    
    // Clear Tokens
    func clearToken() -> Bool {
        let accessTokenConfig = StringSettingConfig(settings: settings, key: "access_token", defaultValue: "")
        let refreshTokenConfig = StringSettingConfig(settings: settings, key: "refresh_token", defaultValue: "")
        
        accessTokenConfig.remove()
        refreshTokenConfig.remove()
        
        return true
    }
}
